### PR TITLE
Security refresh token implementation and oauth2 login

### PR DESCRIPTION
**Pull request by LunguIon**

### CODE REVIEW

File: src/main/java/md/forum/forum/controllers/AuthenticationController.java [hunk 1]
```diff
@@ -1,36 +1,67 @@
...
+        Authentication authentication = authenticationManager.authenticate(new UsernamePasswordAuthenticationToken(loginUserDto.getEmail(), loginUserDto.getPassword()));
+        if (authentication.isAuthenticated()) {
+            RefreshToken refreshToken = refreshTokenService.createRefreshToken(loginUserDto.getEmail());
+            return JwtResponseDTO.builder()
+                    .accessToken(jwtService.generateToken(loginUserDto.getEmail()))
+                    .token(refreshToken.getToken())
+                    .build();
+        } else {
+            throw new UsernameNotFoundException("Invalid user request !");
+        }
```
Comment (InCreating) [inline src/main/java/md/forum/forum/controllers/AuthenticationController.java]: Please, move this logic to service

File: src/main/java/md/forum/forum/controllers/AuthenticationController.java [hunk 2]
```diff
@@ -1,36 +1,67 @@
...
+    public JwtResponseDTO refreshToken(@RequestBody RefreshTokenRequestDTO refreshTokenRequestDTO) {
+        return refreshTokenService.findByToken(refreshTokenRequestDTO.getToken())
+                .map(refreshTokenService::verifyExpiration)
+                .map(RefreshToken::getUser)
+                .map(user -> {
+                    String accessToken = jwtService.generateToken(user.getEmail());
+                    return JwtResponseDTO.builder()
+                            .accessToken(accessToken)
+                            .token(refreshTokenRequestDTO.getToken()).build();
+                }).orElseThrow(() -> new RuntimeException("RefreshToken is not in the DB !"));
```
Comment (InCreating) [inline src/main/java/md/forum/forum/controllers/AuthenticationController.java]: Please, move this logic to service

File: src/main/java/md/forum/forum/controllers/CommentController.java [hunk 3]
```diff
@@ -2,55 +2,97 @@
...
+        logger.info("getAllComments returned {} comments", comments.size());
         return ResponseEntity.ok(comments);
     }
 
+    @GetMapping("/post/{postId}")
+    public ResponseEntity<List<Comment>> getCommentsByPostId(@PathVariable Long postId) {
+        logger.info("getCommentsByPostId was called");
+        List<Comment> comments = commentService.getCommentsByPostId(postId);
+        logger.info("getCommentsByPostId returned {} comments", comments.size());
+        return ResponseEntity.ok(comments);
```
Comment (InCreating) [inline src/main/java/md/forum/forum/controllers/CommentController.java:37]: Please add logging to every new method in controllers

File: src/main/java/md/forum/forum/ForumApplication.java [hunk 4]
```diff
@@ -1,19 +1,16 @@
...
 
-	static Logger logger = Logger.getLogger(ForumApplication.class.getName());
-
-	public static void main(String[] args) {
-		logger.info("App starts");
-		SpringApplication.run(ForumApplication.class, args);
-	}
-
+    public static void main(String[] args) {
+        logger.info("\n\n\n      NEW SESSION\n");
```
Comment (InCreating) [inline src/main/java/md/forum/forum/ForumApplication.java:13]: Why do we need this?

File: logs/forum.log
Comment (InCreating) [inline logs/forum.log:1]: Please remove this file.

File: src/main/java/md/forum/forum/controllers/LikeController.java [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+    // Endpoint to find all likes by a specific user
+    @GetMapping("/user/{userEmail}")
+    public ResponseEntity<List<Like>> findAllByUser(@PathVariable("userEmail") String userEmail) {
+        logger.info("Find all likes by user email: {}", userEmail);
+        List<Like> likes = likeService.findAllByUser(userEmail);
+        logger.info("Found {} likes by user email: {}", likes.size(), userEmail);
+        if (!likes.isEmpty()) {
+            return ResponseEntity.ok(likes);
+        } else {
+            return ResponseEntity.noContent().build();
```
Comment (InCreating) [inline src/main/java/md/forum/forum/controllers/LikeController.java:49]: Why do we need this? Is there any validation for no content status on frontend? If not, please remove this and return "status.Ok" even with empty response. Please fix for other places.

File: src/main/java/md/forum/forum/controllers/PostController.java [hunk 6]
```diff
@@ -11,50 +13,81 @@
...
+        logger.info("deletePost was called with id: {}", id);
         boolean deleted = postService.deletePost(id);
-        return deleted ? ResponseEntity.noContent().build() : ResponseEntity.notFound().build();
+        if (deleted) {
+            logger.info("Post was deleted with id: {}", id);
+            return ResponseEntity.noContent().build();
+        } else {
+            logger.error("Post was not found, and was not deleted");
+            return ResponseEntity.notFound().build();
+        }
```
Comment (InCreating) [inline src/main/java/md/forum/forum/controllers/PostController.java:91]: Why do we need this? Is there any validation for no content status on frontend? If not, please remove this and return "status.Ok" even with empty response. Please fix for other places. We don't even need to check if the entity was really deleted.

File: src/main/java/md/forum/forum/controllers/UserController.java [hunk 7]
```diff
@@ -2,75 +2,131 @@
...
+            return ResponseEntity.status(HttpStatus.CONFLICT).build();
+        } else {
+            logger.info("User created: {}", createdUser.getEmail());
+            return ResponseEntity.status(HttpStatus.CREATED).body(createdUser);
+        }
+
+
     }
 
+    //don't use it !!!!!! for frontend
```
Comment (InCreating) [inline src/main/java/md/forum/forum/controllers/UserController.java:60]: What does this mean? Please explain

File: src/main/java/md/forum/forum/models/Like.java [hunk 8]
```diff
@@ -21,15 +21,14 @@ public class Like {
     private boolean upvote;
     @Column(name = "create_date")
     private Date createDate;
-    @JoinColumn(name = "post_id",foreignKey = @ForeignKey(name = "fk_like_post"))
+    @JoinColumn(name = "post_id", foreignKey = @ForeignKey(name = "fk_like_post"))
```
Comment (InCreating) [inline src/main/java/md/forum/forum/models/Like.java:24]: Please, do not make additional refactor if it is not related to your ticket. We can raise separate ticket for refactor.

File: src/main/java/md/forum/forum/services/LikeService.java [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+    private final UserRepository userRepository;
+    private final PostRepository postRepository;
+    private final CommentRepository commentRepository;
+
+    public LikeService(LikeRepository likeRepository, UserRepository userRepository, PostRepository postRepository, CommentRepository commentRepository) {
+        this.likeRepository = likeRepository;
+        this.userRepository = userRepository;
+        this.postRepository = postRepository;
+        this.commentRepository = commentRepository;
+    }
```
Comment (InCreating) [inline src/main/java/md/forum/forum/services/LikeService.java:28]: Please replace with `@RequiredArgsConstructor`

**Review by InCreating — CHANGES_REQUESTED** [review]

File: src/main/java/md/forum/forum/security/service/RefreshTokenService.java [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+
+    final UserRepository userRepository;
+
+    public RefreshTokenService(RefreshTokenRepository refreshTokenRepository, UserRepository userRepository) {
+        this.refreshTokenRepository = refreshTokenRepository;
+        this.userRepository = userRepository;
+    }
+
+    public RefreshToken createRefreshToken(String username) {
+        Optional<User> userOptional = userRepository.findByEmail(username);
```
Comment (InCreating) [inline src/main/java/md/forum/forum/security/service/RefreshTokenService.java]: Please rename argument to `email`.

File: logs/forum.log
Comment (LunguIon) [reply logs/forum.log:1]: The file is removed, it doesnt exist on remote

File: src/main/java/md/forum/forum/controllers/LikeController.java (hunk 5 above)
Comment (LunguIon) [reply src/main/java/md/forum/forum/controllers/LikeController.java:49]: @Liebe001 can you please answer the question, and clarify the situation.

File: src/main/java/md/forum/forum/controllers/PostController.java (hunk 6 above)
Comment (LunguIon) [reply src/main/java/md/forum/forum/controllers/PostController.java:91]: @Liebe001 can you please answer the question, and clarify the situation.

File: src/main/java/md/forum/forum/controllers/UserController.java (hunk 7 above)
Comment (LunguIon) [reply src/main/java/md/forum/forum/controllers/UserController.java:60]: that mean that the specific method is only for frontend

File: src/main/java/md/forum/forum/services/LikeService.java (hunk 9 above)
Comment (LunguIon) [reply src/main/java/md/forum/forum/services/LikeService.java:28]: This is not a mistake, if @Liebe001 choose this method of initialization is his decision, this class doesn't have a lot of lines so I don't see a problem, @Liebe001 can you please comment on your decision ? Thx.

File: src/main/java/md/forum/forum/security/service/RefreshTokenService.java (hunk 10 above)
Comment (LunguIon) [reply src/main/java/md/forum/forum/security/service/RefreshTokenService.java]: I'll do that, although is not a mistake, but to be more clear, I'll refactor

File: src/main/java/md/forum/forum/controllers/AuthenticationController.java (hunk 2 above)
Comment (LunguIon) [reply src/main/java/md/forum/forum/controllers/AuthenticationController.java]: Please check last commits. Thanks.

File: src/main/java/md/forum/forum/controllers/AuthenticationController.java (hunk 1 above)
Comment (LunguIon) [reply src/main/java/md/forum/forum/controllers/AuthenticationController.java]: Please check last commits. Thanks.

File: src/main/java/md/forum/forum/controllers/CommentController.java (hunk 3 above)
Comment (LunguIon) [reply src/main/java/md/forum/forum/controllers/CommentController.java:37]: @Liebe001  can you please comment on this matter. Thanks.

File: src/main/java/md/forum/forum/ForumApplication.java (hunk 4 above)
Comment (Liebe001) [reply src/main/java/md/forum/forum/ForumApplication.java:13]: image
Just to find old issues faster

File: src/main/java/md/forum/forum/controllers/LikeController.java (hunk 5 above)
Comment (Liebe001) [reply src/main/java/md/forum/forum/controllers/LikeController.java:49]: it's made for logs, my bad, i forgot to add logs, it should look like this
image

File: src/main/java/md/forum/forum/controllers/PostController.java (hunk 6 above)
Comment (Liebe001) [reply src/main/java/md/forum/forum/controllers/PostController.java:91]: i understand that we don't need to check if entity was deleted but it's made for detailed logging

**Comment by LunguIon** [comment]
I'll open another pull request with all changes implemented requested by colleagues

**Pull request closed without merging**